### DISCUSSION TITLE
Any idea for a replacement of gervill?

### DISCUSSION THREAD

**Discussion by guiv42** [helge17/tuxguitar#185] [General]
Some warnings are raised during compilation of [gervill](https://github.com/helge17/tuxguitar/tree/master/desktop/gervill): use of deprecated APIs.
The [url of gervill website in readme](https://github.com/helge17/tuxguitar/blob/a80ea4374241b47c5a5a114b8ce8c2ffa5474879/README.md?plain=1#L19) is no more accessible, so it seems this dependency is no more maintained.
Any suggestion for a possible replacement?

**Comment by harborsiem** [comment]
Gervill is an integrated feature in JDK7 and above. [See](https://docs.oracle.com/javase/8/docs/technotes/guides/sound/enhancements_7.html).
I think that you have to do only simple changes to use gervill from the JDK.

**Reply by guiv42** [reply]
Unfortunately it does not seem as simple as that. I tried to find SoftSynthesizer class (used by org.herac.tuxguitar.midi.synth.impl.GervillProcessor), and I cannot find it.
Could it be linked to [this question](https://stackoverflow.com/questions/70643111/why-cant-i-import-from-com-sun-media-sound)?

**Comment by harborsiem** [comment]
If you don't have access to com/sun/media/sound classes and interfaces (SoftSynthesizer, AudioSynthesizer, EmergencySoundbank, ModelPatch, SF2Instrument, ...) than I think, that Gervill is not well integrated to the Java sound.
The only thing you can do is to ask at the java development team for access changes for Gervill classes and interfaces.
 In my eyes most of the Gervill code is useless.

**Reply by guiv42** [reply]
Thanks for your help.
>  I think, that Gervill is not well integrated to the Java sound

That's my understanding: Gervil *implementation* was integrated, but its APIs are not published. So from our perspective it's just as if it vanished.

> In my eyes most of the Gervill code is useless

Does that mean you think an implementation on top of pure java sound API is possible? For sure it would me more future-proof. I had a very quick look at that code, and I don't know if this is feasible. Just 2 examples (detailed analysis still to be done):
- [SoftSynthesizer instantiation](https://github.com/helge17/tuxguitar/blob/ec25fa455b636d2743d57152bf943bfb25b4292a/desktop/TuxGuitar-synth-gervill/src/org/herac/tuxguitar/midi/synth/impl/GervillProcessor.java#L41C8-L41C8). Java [Synthesizer](https://docs.oracle.com/javase/8/docs/api/javax/sound/midi/Synthesizer.html) is an interface and cannot be instantiated. Would `MidiSystem.getSynthesizer()` be an acceptable replacement? No idea.
- [creation of a stream](https://github.com/helge17/tuxguitar/blob/ec25fa455b636d2743d57152bf943bfb25b4292a/desktop/TuxGuitar-synth-gervill/src/org/herac/tuxguitar/midi/synth/impl/GervillProcessor.java#L99C7-L99C7): I could not find any alternative to replace `AudioSynthesizer.openStream` method

**Comment by harborsiem** [comment]
I had a deeper look to the java sound system and I think I have an idea how to use Gervill with the SoftSynthesizer class.

In com/sun/media/sound there is a class SoftProvider which implements Service Provider Interface (SPI) MidiDeviceProvider. SoftProvider do a SoftSynthesizer instantiation by calling MidiSystem.getMidiDevice(MidiDevice.Info). It returns interface MidiDevice which you can cast to interface Synthesizer. Before calling MidiSystem.getMidiDevice you have to call MidiSystem.getMidiDeviceInfo(). From the result of this call you have to extract the suitable MidiDevice.Info.
You have to replace openStream method and some other things. But this should be the first step to use the gervill synthesizer.

In Tuxguitar code there are already parts that uses Gervill ([TuxGuitar-gervill](https://github.com/helge17/tuxguitar/tree/master/desktop/TuxGuitar-gervill), Gervill as MidiPort in TuxGuitar-jsa and maybe the export of Audio file).

**Reply by guiv42** [reply]
`openStream` seems more of a problem.
It ends up building an AudioInputStream from a custom implementation of an [InputStream](https://github.com/helge17/tuxguitar/blob/2ec4f9762a032e28c104d55eccd8d6577e5af417/desktop/gervill/src/main/java/media/sound/SoftMainMixer.java#L861) with overridden methods. Sounds a bit complex for me, I don't understand much there.
Note: this code is also duplicated in [Android app](https://github.com/helge17/tuxguitar/tree/master/android/TuxGuitar-android-gervill)

**Reply by harborsiem** [reply]
Instead of using an old Gervill implementation in the GervillProcessor you should change it to the new code with usage of the Java sound lib. You can use openStream like before. Then you don`t need the extra Gervill code.

Actual code:
```
	public GervillProcessor(TGContext context) {
		this.context = context;
		this.synth = new SoftSynthesizer();
		this.program = new GervillProgram();
		this.soundbankFactory = new GervillSoundbankFactory();
		this.buffer = new byte[TGAudioBuffer.CHANNELS * TGAudioBuffer.BUFFER_SIZE];
		this.outputs = new float[TGAudioBuffer.CHANNELS][TGAudioBuffer.BUFFER_SIZE / 2];
		for(int i = 0; i < this.outputs.length; i++) {
			this.outputs[i] = new float[TGAudioBuffer.BUFFER_SIZE / 2];
		}
		this.program.setProgram(-1);
	}
```

New code:
```
	public GervillProcessor(TGContext context) {
		this.context = context;

AudioSynthesizer synthesizer = null;
MidiDevice.Info[] infos = MidiSystem.getMidiDeviceInfo();
for (int i = 0; i < infos.length; i++)
{
    if (infos[i].getName() == "Gervill")
    {
        MidiDevice device = MidiSystem.getMidiDevice(infos[i]);
        synthesizer = (AudioSynthesizer)device;
        break;
    }
}
this.synth = synthesizer;

		this.program = new GervillProgram();
		this.soundbankFactory = new GervillSoundbankFactory();
		this.buffer = new byte[TGAudioBuffer.CHANNELS * TGAudioBuffer.BUFFER_SIZE];
		this.outputs = new float[TGAudioBuffer.CHANNELS][TGAudioBuffer.BUFFER_SIZE / 2];
		for(int i = 0; i < this.outputs.length; i++) {
			this.outputs[i] = new float[TGAudioBuffer.BUFFER_SIZE / 2];
		}
		this.program.setProgram(-1);
	}
```

**Reply by harborsiem** [reply]
[GervillProcessor.java](https://github.com/user-attachments/files/23148011/GervillProcessor.java)

An other way is to use reflection for calling openStream method.

**Reply by harborsiem** [reply]
See Oracle Java bug report for interface [AudioSynthesizer](https://bugs.java.com/bugdatabase/view_bug?bug_id=6938913).
Nothing happens !

**Comment by guiv42** [comment]
Thank you @harborsiem to re-open this topic, we may get a serious issue of software obsolescence if it is not handed correctly.
About your proposals:
- new instantiation of `AudioSynthesizer` object: could you please provide a branch with this implemented? I do not understand your proposal, since it still uses `AudioSynthesizer` class, that is defined in Gervill module currently
- use of reflection: I don't think it's a good idea. In practice it does not solve the obsolescence issue, since there is zero guarantee that internal methods will remain unchanged in the future. And what makes it worse: this kind of deprecation cannot even be detected at compile time, so it just makes the app fail at runtime. TuxGuitar already had an issue like this when an internal method that was used by reflection was deprecated (see [this post](https://github.com/helge17/tuxguitar/issues/355#issuecomment-2458045852))

**Comment by helge17** [comment]
FYI: I tried to unify the Gervill code for desktop and android and bring it a little closer to [OpenJDK](https://github.com/openjdk/jdk/tree/92e1357dfd2d874ef1a62ddd69c86a7bb189c6a2/src/java.desktop/share/classes/com/sun/media/sound) with the commits aa7b19ac5f4471a1bd3147324a3b61aae0f131ad, 96c05ab7ee23d9374a666edee49a38725f9a2038 and c821337a2953dfb08eaff65412542ad569f9b063. The code for [desktop](https://github.com/helge17/tuxguitar/tree/96c05ab7ee23d9374a666edee49a38725f9a2038/desktop/gervill/src/main/java/media/sound) and [android](https://github.com/helge17/tuxguitar/tree/96c05ab7ee23d9374a666edee49a38725f9a2038/android/TuxGuitar-android-gervill/src/main/java/media/sound) is now quite the same, but still far away from OpenJDK. However, the task was getting too complicated and time-consuming. I thought the commits might still be a small improvement.
